### ISSUE TITLE
Adventure game in Logo

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#1268]
By @SimHacker:  
https://donhopkins.com/home/archive/logo/adventure.logo

> distributed with Terrapin C64 Logo. They wanted a programming example that was more about data structures and list processing, than turtle graphics and geometry.

Maybe something to get running with our Logo implementations?

**Comment by Alexey-Slyusar** [comment]
Oh, a just wonderful logo code!

It's something difficult to run the code in BBN LOGO or CLOGO because the "sentence" data type that both versions can handle is just kind of flat lists data type without nested list allowed. So code must be heavily rewritten for BBN LOGO and CLOGO. 
 
But now, we have LLISP logo running and this version has full featured list processing capabilities. I'll try to run the code on LLISP. This will be interesting by itself and in addition such beautiful code is very suitable to test LLISP logo version that we have on ITS.

**Comment by Alexey-Slyusar** [comment]
I am well noted that the code is for Terrapin C64 Logo, but I think that I can run it in LLOGO with only minor changes.

**Comment by Alexey-Slyusar** [comment]
As the first step, to feel the code, I've adapted the code to run it on Berkeley Logo:

![default](https://user-images.githubusercontent.com/42380964/46227282-68142a80-c367-11e8-891b-2e0b9319147c.png)

**Comment by Alexey-Slyusar** [comment]
I enjoyed a lot both the code and game by itself.
@SimHacker thank you very much for such great experience.

**Comment by Alexey-Slyusar** [comment]
The program uses DEFINE procedure to create the auxiliary procedures on the fly, during it execution. It is very good from the elegant design point of view, but it is the main problem for porting.
I haven't found the similar procedure in LLOGO yet.
But at any case, I believe that it is possible to write such procedure in MACLISP and extend LLOGO vocabulary. And it may be interesting project by itself :-)

**Comment by Alexey-Slyusar** [comment]
Due to the fantastic LLOGO ability to call the MACLISP directly from LLOGO command prompt and use the MACLISP code in LLOGO programs in mixed LLOGO/MACLISP syntax manner, there is no necessarily to extend LLOGO vocabulary. The only need is to express the same idea as DEFINE but in MACLISP, so I am carefully studying the Pitmanual to be able to do it.

**Comment by larsbrinkhoff** [comment]
Is there any similarity between Terrapin C64 Logo, and Terrapin Apple II Logo?  I'm asking because MIT licensed their Apple II Logo to Terrapin, and what we have in #934 is an early version of that.

LLOGO uses DEFINE internally to define primitives.  Maybe it's the same idea as Terrapin C64 Logo's DEFINE?

**Comment by Alexey-Slyusar** [comment]
I've downloaded Terrapin C64 Logo, and Terrapin Apple II Logo documentation and check the things against each other. It seems that the both versions have very similar syntax. 
More than that, you posted the link in #620 to very interesting discussion at comp.lang.logo. The author of the post at comp.lang.logo is one of the key developer both Terrapin C64 Logo, and Terrapin Apple II Logo and that's what he said: 
"I had to convert Logo from 6502 code to 6809, and remove every reference to the empty list [] because reading or writing that location made the C64 disk go haywire!  I studied secret C64 kernel listings printed in blue on green paper (copy protection) to figure out which Page 0 locations we could use." So Terrapin C64 Logo is direct Terrapin Apple II Logo descendant, and it seems that there are only differences between the versions mostly related to 6502 and 6809 processors differences.

Thank you for the comp.lang.logo link, Lars! The group seems not to be so active now, but nevertheless it is amazing source for interesting logo related information!

And thanks for the hint! I am going to study whether LLOGO internal DEFINE using approach can be applied for porting purpose.

**Comment by larsbrinkhoff** [comment]
Here is Spacewar for "Apple Logo", I'm guessing the LCSI version.

https://paleotronic.com/2018/09/27/ancient-tongues-logo-spacewar/

CC @pigulon

**Comment by Alexey-Slyusar** [comment]
Lars, it is real treasure - the game, the code, the article!!!
The very existance of the Spacewar! Logo code emphasises the Logo culture
origins!
I can't  express my appreciation in words, only by the code:
REPEAT 1000000 [Thank you so much!]


5 окт. 2018 г. 11:29 пользователь "Lars Brinkhoff" <notifications@github.com>
написал:

Here is Spacewar for "Apple Logo", I'm guessing the LCSI version.

https://paleotronic.com/2018/09/27/ancient-tongues-logo-spacewar/

CC @pigulon <https://github.com/pigulon>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/PDP-10/its/issues/1268#issuecomment-427287060>, or mute
the thread

**Comment by Alexey-Slyusar** [comment]
I'll defenitely adapt the code to one of the modern Logo version or run the
original code on Apple II Logo under an emulator. There is a note in the
article that code runs to slowly on Apple II Logo, so emulator must be
running in fast mode instead exact emulation mode.
I can't wait to see the code running.

5 окт. 2018 г. 11:29 пользователь "Lars Brinkhoff" <notifications@github.com>
написал:

Here is Spacewar for "Apple Logo", I'm guessing the LCSI version.

https://paleotronic.com/2018/09/27/ancient-tongues-logo-spacewar/

CC @pigulon <https://github.com/pigulon>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/PDP-10/its/issues/1268#issuecomment-427287060>, or mute
the thread

**Comment by Alexey-Slyusar** [comment]
Here is The Rolling Stone magazine 7 December 1972 article titled 
"S P A C E W A R Fanatic Life and Symbolic Death Among the Computer Bums". 

http://www.wheels.org/spacewar/stone/rolling_stone.html

The article is about The first video game tournament was held on October 19th, 1972. Competitors gathered at the Stanford Artificial Intelligence Lab in Los Altos, California to do battle in the Spacewar.

Please notice the Appendix Two: "Your Own Spacewar" where the Alan Kay's Spacewar version in Logo is posted. Thus, the Logo Spacewar tradition has a long time history. :-)

**Comment by Alexey-Slyusar** [comment]
I've got the Adventure game running on ITS :-)

![default](https://user-images.githubusercontent.com/42380964/46626903-67d31680-cb41-11e8-9d1a-26dd582d3a83.png)

It is the first successful run, so I am continuing to adapt the program for LLOGO.

**Comment by Alexey-Slyusar** [comment]
I've created the `alexey/adventure` branch and put the program source file into the `src/llogo` directory.
@larsbrinkhoff do you think that it is good place for the source file?

**Comment by larsbrinkhoff** [comment]
I think it should be in Don's home directory A2DEH.

**Comment by SimHacker** [comment]
Wow, that’s wonderful! Like having an old friend back from the dead. Or visiting the house you grew up in! 

Yeah that function definition stuff was pretty crazy! I must have been trying to show of Logo’s essential lispyness! But making functions that have side effects is considered quite un-lisp-like. But some times you just gotta get stuff done. 

-Don

**Comment by Alexey-Slyusar** [comment]
Hi Lars! Thanks for the link on some more background information.

**Comment by Alexey-Slyusar** [comment]
@larsbrinkhoff could you please say where A2DEH directory is located in the branch? Thank you!

**Comment by larsbrinkhoff** [comment]
It's not there yet, so you'll have to make it.  `mkdir src/a2deh`, and add `a2deh to SRC in Makefile.

**Comment by Alexey-Slyusar** [comment]
Thanks, I'll make it tonight.

вт, 9 окт. 2018 г. в 8:51, Lars Brinkhoff <notifications@github.com>:

**Comment by Alexey-Slyusar** [comment]
I know that the Adventure game code is not fully functional. But it has
interesting functional programming elements.
Logo world has purely functional code examples such Albert Cuoco's
Investigations in Algebra (I read the book in short parts that available
from Google Books, but I believe that it is true for all book's code. I've
tried to get the whole book in hard copy but unsuccessful - international
book delivery in Amazon is just terrible). And some examples in Brian
Harvey's wonderful CSLS trilogy. (Which I am continuing to learn).

And am I right that the LISP LOGO implementation written in not fully
functional way?


<aleksey.slyusar@gmail.com>:

**Comment by SimHacker** [comment]
There are several meaning of "functional", and the Adventure game might be a good way to help explain that. 
One sense is simply that something works, and is not broken. 
Another sense is that something manipulates functions, which it does. 
A deeper sense is that it manipulates functions in a particularly constrained way, without side effects, which it doesn't. 
It uses a mutable model of the world. The rooms and connections between them don't change, but you could add mud-like authoring and editing commands like "dig Room", "link Room Direction" to edit the room model.
You edit the item model when you drop or pick up an item, by mutating its room number. 
Another aspect of the model is the player's room number, which you change by moving from room to room. 
And yet another aspect of the model is the value of the pronoun "it", which gets set to the last item you named, because the name is actually a function that sets "it" to the item name and returns the item same name. That's not good functional oriented programming style, but it gets the job done. 
Logo lets you program in a mix of procedural and functional styles. 
Off the top of my head I can't think of a way to implement a purely functional adventure game! 
But it might be like heaven, which is a place where nothing ever happens.

**Comment by Alexey-Slyusar** [comment]
@SimHacker thank you for the explanation! The explanation really helps to understand things better! To be honest I am still thinking about the sense of the explanation. And I think that I will recall the explanation repeatedly. Thanks again!

**Comment by Alexey-Slyusar** [comment]
I get this error in LLOGO time to time :
```

;TOO MANY RECURSIONS. USED 4982 WORDS. CONTINUE ANYWAY? 
<y
; OK.

;UNRECOVERABLE REG PDL OVERFLOW
MPV; 11532>>HRRZM 11,(10)   
```
I realise that LLOGO teaches me to use machine memory resources very carefully :-)

@larsbrinkhoff @eswenson1  could you please explain what is 4982 words mean? What kbytes is it? 

And in relation to disk storage allocation AIM-307a says:
![default](https://user-images.githubusercontent.com/42380964/46849104-726a0600-cdf6-11e8-858a-975230ab68e5.png)

How can I translate 26 blocks (1024 36 bits word) to nowadays bytes?

And to make things clear enough what is PDP-10 processor frequency? As far as I am understand PDP-10 is 36-bit architecture machine so it is difficult to compare performance with other architectures.  Maybe it is not so informative to measure PDP-10 performance in Hz.
Are FLOPS or MIPS more suitable measurement than Hz?

**Comment by larsbrinkhoff** [comment]
One word is 36 bits.  4982 words are 19928 (packed four per word), or 24910 ASCII characters.  On the stack, one recursion level would normally take at least one word, the return address.  Maybe LLOGO doesn't optimize tail calls, or your code prevents tail calls.

26 blocks would be something like 104K octets.

The KA10 processor is asynchronous, so there is no clock speed.  There is memor cycle times, which are on the order of 1 microsecond.

**Comment by Alexey-Slyusar** [comment]
Hi Lars!
Thank you very much!

пт, 12 окт. 2018 г. в 17:55, Lars Brinkhoff <notifications@github.com>:

**Comment by larsbrinkhoff** [comment]
Someone else had the same idea!

https://paleotronic.com/2018/11/10/logo-adventure-for-apple-logo/

I don't know who wrote the article.  Maybe @pigulon?

**Comment by pigulon** [comment]
@larsbrinkhoff  I can't take the credit for this one.  My partner ported it to Apple Logo yesterday and wrote the article. I may have consulted a little :)

**Comment by SimHacker** [comment]
@pigulon thanks to you and your partner! You've done such great work together. 

The whole Paleotronic web site is wonderful! It really explained the nitty gritty of Apple ][ hires graphics. 

My first BASIC program was a Lunar Lander game. My zeroth BASIC program just converted between Celsius and Fahrenheit, and somebody helped me, so it didn’t count, or even integrate over time. 

Earlier, I'd learned the “lunar lander algorithm” by playing with one of those TI programmable calculators with the magnetic strip readers that go bzzzzzzt, which had a library of games including lunar lander. 

I just discovered and downloaded microM8, which is a whole lot more than what it modestly describes itself as an "Apple II Emulator for Windows, macOS / OSX and Linux". 

https://paleotronic.com/software/microm8/

It does so much and has so many layers, I'm having a hard time getting my head around it, or knowing where to start. 

I ran it and got some of the logo turtle stuff to work — space wars is the perfect game for demonstrating turtle graphics and teaching physics. 

What's the story behind miccrom8, your favorite demos, what do I do to see 3d pixels, and are there any video demos online? 

I'd love to hear what kind of approach you would you take to a PDP-10 emulator, too!

**Comment by larsbrinkhoff** [comment]
Thanks for the info @pigulon.  I see you used LCSI's Apple II Logo.

We have MIT's Apple II Logo built from source code here:  
http://lars.nocrew.org/its/aplogo/  
https://github.com/PDP-10/its/tree/master/src/aplogo/  
https://github.com/PDP-10/its/tree/master/doc/aplogo/

I wonder how to
1. get a source file such as Hopkins' game onto an Apple II disk image, and
2. how to make Logo load it?

**Comment by SimHacker** [comment]
@pigulon, would your Apple ][ disk management utility diskM8 do the trick, or does it have to be stuffed into the logo interpreter through the keyboard?

https://paleotronic.com/diskm8/

You're so modest about it, but there's some really aM8zing stuff on your web site and in your emulator! ;)

**Comment by pigulon** [comment]
Assuming the file format needed is text (I think it is with logo), then you probably could use diskm8.  
Just Name the file with the txt extension, mount the volume in diskm8, and do "put <localfilename>".  
There's online help in the utility.  Be aware that it's pretty alpha, and may eat your hamster :)  Be sure to make sure that the line endings are correct for the system (diskm8 currently doesn't do anything in this regard).

**Comment by Alexey-Slyusar** [comment]
I'm working to get the same Don's Adventure game look'n'feel on LLOGO as on original C64 Terrapin C64 Logo.

In addition I'm working on following programs implemented in LLOGO:
* A BASIC Compiler;
* A LLOGO version for Eliza program  written in 1966 by Joseph Weizenbaum;
* a Logo version of Student, a program written by Daniel G. Bobrow for his 1964 Ph.D. thesis, Natural Language Input for a Computer Problem Solving System, at MIT;

All these great Logo code written by Dr. Brian Harvey for his awesome CSLS trilogy. 
I just adapting the code for LLOGO to see clearly possibilities and limitations for our current LLOGO realisation and make some concerns about possible further Lisp Logo enchantments.

I made the `alexey/adventure` branch and plan to upload working LLOGO versions of the programs into the branch.

But there is a question arises.

Possibly it isn’t clear that all that code (including the Adventure game) belongs in the ITS repository – since it was not historically part of ITS and is really unrelated to ITS. 
Maybe it would be better to create a separate Github repository for it. But I don't want to do that actually.

As a compromise, I think to keep the `alexey/adventure` branch here, but not to merge it with the master branch. Is it good idea?
Any the question related suggestions from PDP-10/ITS community members are highly appreciated.

**Comment by larsbrinkhoff** [comment]
This repository serves both to preserve ITS and also update it with bug fixes and new features.

We have added a small number of programs that never ran on ITS.  Since a few additional Logo program in no way does any harm to the existing software, I think it would be OK to add them to the master branch.  I think this is in the spirit of ITS, where the user brought in programs from outside MIT and ported them to ITS.

**Comment by Alexey-Slyusar** [comment]
Thank you for the comment, Lars! I very appreciate it. It makes things clear.

I've got the Lisp Logo adventure version look'n'feel exactly as the original one. With such thing like:
"??VE ME AND  YOU'LL BE GLAD" as on the original version screenshot:
![originaladventurescreen3](https://user-images.githubusercontent.com/42380964/50636424-103d1f00-0f67-11e9-9ea2-c598f84707b1.JPG)
and
"YOU DON'T HAVE THE PRPPER TOOLS TO FIX THAT"
![originaladventurescreen2](https://user-images.githubusercontent.com/42380964/50636178-28606e80-0f66-11e9-86ba-a7c501dbd7d4.JPG)

and endless loop in `DONE` definition. So ^G is required to stop invoking DONE at the end of the game.

By the way, @SimHacker could you, please, say is `PRPPER TOOL` phrase on the second screenshot is typo or it is some kind of jargon phrase?

Among solving other syntactic differences it was not so straightforward to find the way to keep all the original punctuation unmodified. Because Lisp Logo, unlike all other logo versions, evaluates words / atoms inside list. So, expression `PRINT [YOU'LL BE GLAD]` evaluates as `YOU (QUOTE LL) BE GLAD` and [FOO. BAR] treated as dotted pair by the interpreter.

I have a compiled version of the game as well. But it is lightly different from the interpretive one and from the original version due to the name conflicts issues, that cannot be avoiding during compilation.
@larsbrinkhoff is it good idea to put the compiled version in the repo or everyone who wanted should better compile one on their own?

**Comment by larsbrinkhoff** [comment]
1. Maybe Maclisp's quoting character `/` works somehow inside Logo lists?  Try `PRINT [FOO/'BAR]`.

2. Not everything have to be compiled.  I think it's OK to have this Adventure be a loadable example.

**Comment by SimHacker** [comment]
Well the code predates my obsession with Parappa the Rapper, so a PRPPER TOOL probably doesn't mean "microphone". 

https://www.youtube.com/watch?v=F5Pm7BL-hyo

When I wrote up the Medium page about the code, I think I corrected a couple of original spelling typos, but I can't remember what they were. But you can use that code as a reference for what I meant. Looks like I corrected PRPPER to PROPPER but kept the double P's so it's still misspelled! ;) Since that's an important clue, you should spell it PROPER. 

https://medium.com/@donhopkins/logo-adventure-for-c64-terrapin-logo-4c684a240b53

TO FIX :ITEM
 IF IHAVE? :ITEM PR [YOU HAVE TO DROP IT TO FIX IT!] CMD
 IF NOT HERE? :ITEM SEENO :ITEM
 IF NOT :ITEM = "MACHINE PR [YOU CAN'T FIX THAT!] CMD
 IF NOT ITEMLOC "WAND = 0 PR [THE MACHINE IS NOT BROKEN!] CMD
 IF NOT IHAVE? "SCREWDRIVER PR [YOU DON'T HAVE THE PROPPER TOOLS TO] PR [FIX IT] CMD
 PR [YOU FIX THE MACHINE WITH YOUR TRUSTY]
 PR [SCREWDRIVER. UPON BEING FIXED, THE]
 PR [MACHINE STARTS UP AND PRODUCES A WAND!]
 PUTITEM "WAND 4
 CMD
END

Hmm, not sure what to do about the quoting problems, but Lars' suggestion about trying / is a good idea. 

Thanks for working on the code and bringing it back to life!

**Comment by SimHacker** [comment]
It puzzles me how this would work, with the uneven number of double quotes:

 IF :ITEM = "WAND PR [IT BEARS A FADED INSCRIPTION:] PR ["WAVE ME AND YOU'LL BE GLAD."] CMD

Is that another typo or transcription error? 

I think I pulled this code directly out of a binary C64 floppy disk image with emacs, so it might have been fragmented and I might have made mistakes reassembling it, or the cat might have walked across the keyboard when I wasn't looking. A lot of bugs can be traced down to that problem.

**Comment by Alexey-Slyusar** [comment]
Conceptually,  quotation marks are not used in pairs in Logo. This is not
just arbitrary syntactic foible; it reflects the fact that a Logo *word*
data type is a different idea  from what would be called *character string*
in other languages.
But in case of Lisp Logo we have the situation is confusing a little,
because it is only Logo version that doesn't fully follow this convention.
I'll post more extensive explanation for the issue as soon as I can, with
my big pleasure I must say :-)

пт, 4 янв. 2019 г. в 10:23, Lars Brinkhoff <notifications@github.com>:

**Comment by larsbrinkhoff** [comment]
Looking forward to it.  Please include BBN LOGO quotation syntax too.

We have some starter guides for various things in ITS here: https://github.com/PDP-10/its#documentation  It would be good to add one for Logo too.

**Comment by Alexey-Slyusar** [comment]
Thank you Lars! Well noted.

пт, 4 янв. 2019 г. в 12:16, Lars Brinkhoff <notifications@github.com>:

**Comment by Alexey-Slyusar** [comment]
When I was reading Lars's post above, where BBN LOGO mentioned, I realised that the logo quoting issue has one more dimension if we taking into account this version too (and CLOGO as its descendant). And the things become a little more complicated to explain.

Well, lets try to arrange the things in right order.

Hopefully, despite all these syntactic differences, Logo is very well designed language with the underlying principles that mostly common for all historical and nowadays Logo realisations. In that aspect Logo is similar with its parent language Lisp, but unlike Lisp, Logo doesn't have a standard even today.

Basically, as in Lisp, in Logo to quote something means to prevent it from being evaluated.

Lets see the idea in concrete example using 11LOGO syntax and semantics (most current Logo versions and Logo versions from 80s (including C64Logo) and 90s follow in general this syntax) and LLOGO syntax. 
There are only 4 ways that a word can appear in Logo, here is 11LOGO syntax:
```
?PRINT "FOO                     FOO is quoted
?PRINT [FOO BAR]            FOO BAR are packaged in a list
?PRINT :FOO                     FOO is dotted
?PRINT FOO                      FOO is unadorned
```
In 11LOGO, the double quote character `"` is used to specify that the atom following it is not to be evaluated:
```
?PRINT "FOO
FOO
```
It is like LISP's single quote, except that it also affects the LOGO reader's decision about when to stop including successive characters in forming the name of an atom.
Following a double quote, the only separator characters recognised are space, carriage return, and square brackets, Thus, in 11LOGO,
```
?PRINT "FOO+3
FOO+3
```
In LLOGO, the user may use the LISP single quote to specify that an atom or parenthesized list following the single quote is not to be evaluated. The presence of the single quote does not change the way LLOGO decides when an atom ends. Because the plus sign is still a separator character, LLOGO uses the double quotes as CLOGO does:
```
?PRINT "FOO+3"
FOO+3
```
Square brackets in 11LOGO specify quoted lists. Square brackets in 11LOGO also share with double quotes the property described above of affecting the LOGO reader's decision on ending the names of atoms. Within a square bracketed list in 11LOGO, an atom is terminated only by a space, carriage return or bracket. This property is not true of square brackets in LLOGO. In LLOGO, [FOO+3] is a list containing three elements, but in 11LOGO, it contains only one element. So, in 11LOGO
```
?PRINT ["WAVE ME AND YOU'LL BE GLAD."] 
"WAVE ME AND YOU'LL BE GLAD."
```
But it is not true for LLOGO. Happily, there is solution for such cases in LLOGO: string quoting in LLOGO is accomplished using the dollar sign character, $. LLOGO will treat anything appearing between dollar signs literally, with special characters devoid of any special meaning. So, in LLOGO
```
?PRINT [$"WAVE ME AND YOU'LL BE GLAD."$] 
"WAVE ME AND YOU'LL BE GLAD."
```
I use that feature extensively to keep original adventure game punctuation in my LLOGO port.

I'll shortly comment things in BBN LOGO and CLOGO in my next post.

**Comment by larsbrinkhoff** [comment]
Thanks, looking forward to the next post in the series. :-)

**Comment by Alexey-Slyusar** [comment]
A pleasure. :-)
Sorry for the delay.
Get back soon.

вт, 8 янв. 2019 г. в 11:43, Lars Brinkhoff <notifications@github.com>:

**Comment by Alexey-Slyusar** [comment]
Before I forget to do this, I want to say:

@SimHacker thank you very much for the comment!

And due to your words: 
> Since that's an important clue, you should spell it PROPER. But maybe APPROPRIATE would be more appropriate.

I've changed PRPPER to APPROPRIATE so It becomes LLOGO port feature sanctioned by the author of the original version.

@larsbrinkhoff wrote:
>We have some starter guides for various things in ITS here: https://github.com/PDP-10/its#documentation It would be good to add one for Logo too.
Good point, Lars, thank you. We have a lot of BBN LOGO / CLOGO / 11LOGO / LISP LOGO related information here, scattered among different issues currently. So it is worth to summarise the information into some kind of starter guide for Logo versions that we have running on ITS. I am going to create such issue and get this task to work on.

**Comment by Alexey-Slyusar** [comment]
Well, lets get back to the question now.

BBN LOGO and CLOGO have the same underlying ideas as above but have a little bit confusing syntax to represents these ideas. One possible reason for this is that both BBN LOGO and its descendant CLOGO are earlier Logo versions than 11LOGO and LISP LOGO and MIT Logo Group was in search of right syntax to represent idea of *word* and *list* data types for beginners (more precisely *sentence* in case of the BBN LOGO / CLOGO).

Thus, in BBN LOGO:
```
?PRINT "FOO"                     FOO is quoted
?PRINT "FOO BAR"             FOO BAR are packaged in a sentence 
?PRINT :FOO:                      FOO is dotted
?PRINT FOO                        FOO is unadorned
```
The confusing point here is not only in fact of similarity between *word*, *sentence* BBN LOGO syntax and *character string* in other languages. 
Lets see an example in BBN LOGO:
```
_PRINT FIRST "FOO BAR"
FOO
_PRINT BUTFIRST "FOO BAR"
BAR
```
Wherein FOO is a *word* and BAR is a *sentence* (list), because FIRST and BUTFIRST are an analogy for CAR and CDR respectively. FIRST returns head of list (which is an atom) and BUTFIRST returns list's tail part (which is a list). But it is not so obvious from the results of the expressions evaluation because both results are visually the same. And we can prove it only by using predicates:
```
_PRINT WORDP FIRST "FOO BAR"
TRUE
_PRINT SENTENCEP BUTFIRST "FOO BAR"
TRUE
``` 

In respect of BBN LOGO and CLOGO I do no know the way how to get this properly printed:
`"WAVE ME AND YOU'LL BE GLAD."`

But it is not so important, I think. The more important thing here is that everyone who is interested can trace the evolution of the Logo syntax here. And not only just trace, but also evaluate that syntax in the interpreter. :-)

At the end, I want to say couple of words in respect of the difference between *sentence* and *list* data types.
Lists have a simple recursive definition. A list is either
1. NIL, the empty list
2. (word1 word2 ...), a sentence of words (= atoms)
3. A list of lists.

This definition is confusing when the person is still having trouble with concept of recursion. BBN LOGO and CLOGO due to pedagogic purposes limits itself to lists built from only the first two of these three clauses. Such lists are called *sentences*.

**Comment by larsbrinkhoff** [comment]
Fixed by #1494.
